### ISSUE TITLE
"fully convolutional" extension of mnist example

### ISSUE THREAD

**Issue by mmmikael** [fchollet/keras#356]
I am trying to convert the mnist cnn example to a "fully convolutional" net in order to apply it efficiently across an image (like a sliding window). I did the following changes:

```python
	# model.add(Flatten())
	# model.add(Dense(32*196, 128))
	## replaced by:
	model.add(Convolution2D(128, 32, 14, 14))
```
and 
```python
	# model.add(Dense(128, nb_classes))
	## replaced by:
	model.add(Convolution2D(nb_classes, 128, 1, 1))
	model.add(Flatten())
```

The training step goes well but when testing on input images different than 28x28 I don't get the expected results. For example, here I created an image slightly bigger to have two classifications:
```python
data = np.zeros([1,1,28, 30])
data[0,0,:,:28] = X_train[0,0,:,:]
 
print('label: %d' % y_train[0])
print(model.predict(data[:,:,:,:28]))
print(model.predict(data))            
```
The output is something like:
```
label: 5
[[ 0.     0.     0.     0.046  0.     0.954  0.     0.     0.     0.   ]]
[[ 0.     0.     0.     0.     0.     0.     0.036  0.001  0.     0.     0.756
   0.207  0.     0.     0.     0.     0.     0.     0.     0.   ]]
```

I would expect to see the numbers of the first output in the second one as well since they should correspond to the same data.

Any idea what I am doing wrong?

Here is a gist of the complete script:
https://gist.github.com/mmmikael/cc895249e7941c2f56b3

**Comment by mthrok** [comment]
Is your test data float32? np.zeros defaults to float64. 

Didn't know that this can take different size image...

**Comment by mmmikael** [comment]
> Is your test data float32? np.zeros defaults to float64.

I tried with float32. Same results...

**Comment by mthrok** [comment]
> I would expect to see the numbers of the first output in the second one as well since they should correspond to the same data.

No. Since you are using `softmax`, the result gets normalized to have `sum==1.0` so if the output size changes, then it normalized in different way. So if you don't want the result to get normalized then maybe try `sigmoid` as the final activation.

**Comment by mmmikael** [comment]
@mthrok if the normalization was the single problem, wouldn't the output be scaled by 0.5?

**Comment by mthrok** [comment]
I got the feeling that the output is correct, it's just different order than you expect. 
Making every two elements into one column it looks kinda similar to the first output. 
```python
[[ 0.     0.     0.     0.     0.     0.     0.036  0.001  0.     0.     0.756  0.207  0.     0.     0.     0.     0.     0.     0.     0.   ]]
# Reorder ↓
[[0.      0.     0.     0.036  0.     0.756  0.     0.     0.     0.   ],
 [0.      0.     0.     0.001  0.     0.207  0.     0.     0.     0.   ]]
# which resembles 
[[ 0.     0.     0.     0.046  0.     0.954  0.     0.     0.     0.   ]]
```
This can be explained as follow. 
When you feed `h28 * w28` and `h28 * w29` image, the data shape after second `Dropout ` is `h1 * w1` but when you feed `h28 * w30` image, it is `h1 * w2`. (more precisely, `#class * c1 * h1 * w2`) 
Since Theano uses column major order, when it is flattened at `Flatten()`, the data order is the mixture of two lines, like `probability of [[class1_line1, class1_line2, class2_line1, class2_line2, ..., classN_line1, classN_line2]]`

**Comment by mthrok** [comment]
maybe your guess is correct. it looks proportional.

**Comment by mmmikael** [comment]
Yes, you were right. Thanks! 

After replacing the ```softmax``` activation by a ```sigmoid``` and reordering the output I am getting the expected results. Also, to be totally consistant, I changed the ```border_mode``` of the first convolution to ```valid```. This way, each window evaluation is not influenced by outside pixels.

I have updated my gist [here](https://gist.github.com/mmmikael/cc895249e7941c2f56b3).
